src/System.Net.Http/src/System/Net/Http/DiagnosticsHandler.cs

```diff
@@ -36,19 +38,107 @@ internal static bool IsEnabled()
+            {
+                if (Activity.Current != null)
+                {
+                    responseTask = SendInstrumentedAsync(request, cancellationToken);
+                }
+                else
+                {
+                    //null Activity.Current means that incoming request was not instrumented
+                    //and there is nothing we can do for outgoing request. 
+                    //Activity events are enabled so we don't send Request/Response events
+                    responseTask = base.SendAsync(request, cancellationToken);
+                }
+            }
+            else 
+            {
+                Guid loggingRequestId = Guid.NewGuid();
+                LogHttpRequest(request, loggingRequestId);
+                responseTask = base.SendAsync(request, cancellationToken);
+                LogHttpResponse(responseTask, loggingRequestId);
+            }
             return responseTask;
         }
 
         #region private
 
-        private static readonly DiagnosticListener s_diagnosticListener =
-            new DiagnosticListener(DiagnosticsHandlerLoggingStrings.DiagnosticListenerName);
+        private async Task<HttpResponseMessage> SendInstrumentedAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            Activity activity = null;
+            //check if user wants THIS request to be instrumented
+            if (s_diagnosticListener.IsEnabled(DiagnosticsHandlerLoggingStrings.ActivityName, request))
+            {
+                activity = new Activity(DiagnosticsHandlerLoggingStrings.ActivityName);
+
+                s_diagnosticListener.StartActivity(activity, new { Request = request });
+
+                //inject headers unless user does not want it to happen
+                //we may pass request to IsEnabled, but not clear how useful it is
+                if (s_diagnosticListener.IsEnabled(DiagnosticsHandlerLoggingStrings.ActivityPropagateName))
+                {
+                    request.Headers.Add(HttpKnownHeaderNames.RequestId, activity.Id);
+                    
+                    //we expect baggage to be empty or contain a few items
+                    bool baggageIsEmpty = true;
+                    foreach (var _ in activity.Baggage)
+                    {
+                        baggageIsEmpty = false;
+                        break;
+                    }
+
+                    if (!baggageIsEmpty)
+                    {
+                        List<string> baggage = new List<string>();
+                        foreach (KeyValuePair<string, string> baggageItem in activity.Baggage)
+                        {
+                            baggage.Add(new NameValueHeaderValue(baggageItem.Key, baggageItem.Value).ToString());
+                        }
+                        request.Headers.Add(HttpKnownHeaderNames.CorrelationContext, baggage);
+                    }
```

Comment:
Why enumerate Baggage twice?  How about replacing this and the baggageIsEmpty stuff above with something like:
```C#
using (IEnumerator<KeyValuePair<string, string>> e = activity.Baggage.GetEnumerator())
{
    if (e.MoveNext())
    {
        var baggage = new List<string>();
        do { baggage.Add(new NameValueHeaderValue(baggageItem.Key, baggageItem.Value).ToString()); }
        while (e.MoveNext());
        request.Headers.Add(HttpKnownHeaderNames.CorrelationContext, baggage));
    }
}
```